rust/src/modbus/detect.rs

```diff
@@ -0,0 +1,1345 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::modbus::ModbusTransaction;
+use lazy_static::lazy_static;
+use regex::Regex;
+use sawp_modbus::{AccessType, CodeCategory, Data, FunctionCode, Message};
+use std::ffi::CStr;
+use std::ops::{Range, RangeInclusive};
+use std::os::raw::{c_char, c_void};
+use std::str::FromStr;
+
+lazy_static! {
+    static ref ACCESS_RE: Regex = Regex::new("^\\s*\"?\\s*access\\s*(read|write)\\s*(discretes|coils|input|holding)?(?:,\\s*address\\s+([<>]?\\d+)(?:<>(\\d+))?(?:,\\s*value\\s+([<>]?\\d+)(?:<>(\\d+))?)?)?\\s*\"?\\s*$").unwrap();
+    static ref FUNC_RE: Regex = Regex::new(
+        "^\\s*\"?\\s*function\\s*(!?[A-z0-9]+)(?:,\\s*subfunction\\s+(\\d+))?\\s*\"?\\s*$"
+    )
+    .unwrap();
+    static ref UNIT_RE: Regex =
+            Regex::new("^\\s*\"?\\s*unit\\s+([<>]?\\d+)(?:<>(\\d+))?(?:,\\s*(.*))?\\s*\"?\\s*$")
+                .unwrap();
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DetectModbus {
+    category: Option<CodeCategory>,
+    function: Option<FunctionCode>,
+    subfunction: Option<u16>,
+    access_type: Option<AccessType>,
+    unit_id: Option<Range<u16>>,
```

Comment:
Interesting, did you see `DetectU32Data` in HTTP2 ?
What are the pros and cons for `Range` ?

Reply:
`DetectU32Data` looks very similar to the way it was implemented in the prior [modbus C module](https://github.com/OISF/suricata/blob/master/src/detect-modbus.h#L50).

Our goal with using `Range` was to use functions like `range.contains()` and make it more idiomatic but in the end we've had to handle all the edge cases to make it compatible with the prior C module (i.e. Rust ranges are half-open).

Should we move to a `DetectModbusValue` style structure?

out-of-scope: If this is used across modules, should it be in a common module?

Reply:
The point would be to have a common module to reduce code duplication...
I think this is not a blocking point and you should go as you feel best